src/libusermode/userhook.cpp

```diff
@@ -0,0 +1,821 @@
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = vad_start
+    };
+
+    addr_t export_header_rva = 0;
+    size_t export_header_size = 0;
+
+    constexpr int MAX_HEADER_BYTES = 1024;   // keep under 1 page
+    uint8_t image[MAX_HEADER_BYTES];
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    bool status = (VMI_SUCCESS == peparse_get_image(vmi, &ctx, MAX_HEADER_BYTES, image));
+    drakvuf_release_vmi(drakvuf);
+
+    if (!status)
+        return nullptr;
+
+    void* optional_header = NULL;
+    uint16_t magic = 0;
+
+    peparse_assign_headers(image, NULL, NULL, &magic, &optional_header, NULL, NULL);
+    export_header_rva = peparse_get_idd_rva(IMAGE_DIRECTORY_ENTRY_EXPORT, &magic, optional_header, NULL, NULL);
+    export_header_size = peparse_get_idd_size(IMAGE_DIRECTORY_ENTRY_EXPORT, &magic, optional_header, NULL, NULL);
+
+    if (export_header_rva >= vad_length)
+    {
+        PRINT_DEBUG("[USERHOOK] Export header RVA is forwarded outside VAD\n");
+        return nullptr;
+    }
+    else if (export_header_size >= vad_length - export_header_rva)
+    {
+        PRINT_DEBUG("[USERHOOK] Export header size is forwarded outside VAD\n");
+        return nullptr;
+    }
+
+    dll_meta.pf_current_addr = vad_start + export_header_rva & ~(VMI_PS_4KB - 1);
+    dll_meta.pf_max_addr = vad_start + export_header_rva + export_header_size;
+
+    if (dll_meta.pf_max_addr & VMI_PS_4KB)
+    {
+        dll_meta.pf_max_addr += VMI_PS_4KB;
+        dll_meta.pf_max_addr = dll_meta.pf_max_addr & ~(VMI_PS_4KB - 1);
+    }
+
+    auto it = plugin->loaded_dlls.emplace(info->regs->cr3, std::vector<dll_t>()).first;
+    it->second.push_back(std::move(dll_meta));
+    return &it->second.back();
+}
+
+static bool make_trap(vmi_instance_t vmi, drakvuf_t drakvuf, drakvuf_trap_info * info, hook_target_entry_t * target, addr_t exec_func)
+{
+    target->pid = info->proc_data.pid;
+
+    drakvuf_trap_t* trap = new drakvuf_trap_t;
+    trap->type = BREAKPOINT;
+    trap->name = target->target_name.c_str();
+    trap->cb = target->callback;
+    trap->data = target;
+
+    // during CoW we need to find all traps placed on the same physical page
+    // that's why we'll manually resolve vaddr and strore paddr under trap->breakpoint.addr
+    addr_t pa;
+
+    if (vmi_pagetable_lookup(vmi, info->regs->cr3, exec_func, &pa) != VMI_SUCCESS)
+    {
+        goto fail;
+    }
+
+    trap->breakpoint.lookup_type = LOOKUP_NONE;
+    trap->breakpoint.addr_type = ADDR_PA;
+    trap->breakpoint.addr = pa;
+
+    if (drakvuf_add_trap(drakvuf, trap))
+    {
+        target->trap = trap;
+        return true;
+    }
+
+fail:
+    PRINT_DEBUG("[USERHOOK] Failed to add trap :(\n");
+    delete trap;
+    return false;
+}
+
+static event_response_t perform_hooking(drakvuf_t drakvuf, drakvuf_trap_info * info, userhook * plugin, dll_t * dll_meta)
+{
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
```

Comment:
I would recommend using the lock_guard mechanism instead: `vmi_lock_guard lg(drakvuf);` and then you don't have to worry about unlocking it before each return, its done for your automatically.